File: src/Solutions/Euler019.hs
module Solutions.Euler019 where
import Control.Monad.State

dayInMonth :: Int -> Int -> Int
dayInMonth month year
    | month `elem` [0, 2, 4, 6, 7, 9, 11] = 31
    | month `elem` [3, 5, 8, 10] = 30
    | year `mod` 4 == 0 = 29
    | otherwise = 28

type Date = (Int, Int, Int, Int)
addSunday, step :: Date -> Date
addSunday (w, m, y, s) = (w, m, y, s + 1)
step (w, m, y, s) = ((w + dayInMonth m y) `mod` 7, m', y', s) where
    (m', y') = if m == 11 then (0, y + 1) else (m + 1, y)
                
-- Monday = 0, Tuesday = 1, ..., Sunday = 6
-- January = 0, ..., December = 11
loop :: State Date Int
loop = get >>= \(weekday, _, year, sundays) -> if year == 2001
        then return sundays
        else when (weekday == 6) (modify addSunday) >> modify step >> loop

euler019 :: IO String
euler019 = return $ show $ evalState loop (1, 0, 1901, 0)

File: rust/src/solutions/euler019.rs
fn day_in_month(month: i32, year: i32) -> i32 {
    if [0, 2, 4, 6, 7, 9, 11].contains(&month) {
        31
    } else if [3, 5, 8, 10].contains(&month) {
        30
    } else if year % 4 == 0 {
        29
    } else {
        28
    }
}

struct Date {
    weekday: i32,
    month: i32,
    year: i32,
}

impl Date {
    fn step(&mut self) {
        self.weekday = (self.weekday + day_in_month(self.month, self.year)) % 7;
        if self.month == 11 {
            self.month = 0;
            self.year += 1;
        } else {
            self.month += 1;
        }
    }
}

fn query(date: &mut Date) -> i32 {
    let mut ans = 0;
    while date.year < 2001 {
        if date.weekday == 6 {
            ans += 1;
        }
        date.step()
    }
    ans
}

pub fn euler019() -> String {
    query(&mut Date {
        weekday: 1,
        month: 0,
        year: 1901,
    })
    .to_string()
}
